src/RevocableAllocator.zig

```diff
@@ -0,0 +1,42 @@
+//! An allocator which can be disabled at runtime.
+//! We use this for allocating at startup and then
+//! disable it to prevent accidental dynamic allocation at runtime.
+
+const std = @import("std");
+const assert = std.debug.assert;
+const mem = std.mem;
+const log = std.log.scoped(.init_allocator);
+
+parent_allocator: mem.Allocator,
+allow_allocation: bool,
+
+const Self = @This();
+
+pub fn init(parent_allocator: mem.Allocator) Self {
+    return .{
+        .parent_allocator = parent_allocator,
+        .allow_allocation = true,
+    };
+}
+
+pub fn deinit(self: *Self) void {
+    self.* = undefined;
+}
+
+pub fn allocator(self: *Self) mem.Allocator {
+    return mem.Allocator.init(self, alloc, resize, free);
+}
+
+fn alloc(self: *Self, len: usize, ptr_align: u29, len_align: u29, ra: usize) error{OutOfMemory}![]u8 {
+    assert(self.allow_allocation);
+    return self.parent_allocator.rawAlloc(len, ptr_align, len_align, ra);
+}
+
+fn resize(self: *Self, buf: []u8, buf_align: u29, new_len: usize, len_align: u29, ra: usize) ?usize {
+    assert(self.allow_allocation);
+    return self.parent_allocator.rawResize(buf, buf_align, new_len, len_align, ra);
+}
+
+fn free(self: *Self, buf: []u8, buf_align: u29, ret_addr: usize) void {
+    return self.parent_allocator.rawFree(buf, buf_align, ret_addr);
```

Comment:
Let's work the positive/negative space both ways, symmetrically, so that we also detect dynamic frees where the allocator is still in static mode.

It seems like we should go through these state transitions:

1. allow allocations only, and only before runtime,
2. disallow allocation and frees during runtime,
3. allow frees only, and only after runtime.